### PR TITLE
CIHA RapidPro Support

### PR DESCRIPTION
**Pull request by SCdF**
Branch for changes needed to prototype / test RapidPro integration

### CODE REVIEW

**Comment by SCdF** [comment]
@benkags FYI you should be able to deploy this now, and use this:

```sh
curl -X GET -H 'Content-Type: application/json' http://admin:pass@localhost:5988/api/v1/hydrate -d '["a7324118-0053-4ce5-8364-e47a0e8f073a", "f7dd9204-1a35-44fe-899f-5226f8caa509"]'
``` 

To hydrate docs.

**Comment by michaelkohn** [comment]
Just to confirm, the hydration will include only the "primary" contact of the parent places, correct?

File: api/src/controllers/hydration.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    const docIds = Array.isArray(req.body) ? req.body : [req.body];
+
+    if (docIds.length === 1) {
+      return lineage.fetchHydratedDoc(docIds[0])
+        .then(hydratedDoc => res.json(hydratedDoc));
+    }
+
+    return db.medic.allDocs({keys: docIds, include_docs: true})
+      .then(result => {
+        const docs = result.rows.map(r => r.doc);
```
Comment (SCdF) [inline api/src/controllers/hydration.js]: TODO: deal with docs that don't exist or otherwise error when querying for

**Comment by SCdF** [comment]
@michaelkohn we should really document these rules somewhere if we haven't before. I don't recall exactly how they work, but pulling from examples, the things that get hydrated are:
 - `doc.contact`, all of those contact's parents, those parents contacts, but not those parents contacts parents
 - `doc.parent`, all those parent's parents, those parent's contacts, but not those parents contacts parents
 - `doc.patient`, all of those patient's parents, but not those parent's contacts (is this a bug @garethbowen ?)

For reference here is a hydrated person:

```json
{
    "_id": "4a97ae47-ebfe-4d39-8f4e-cf65766f6b80",
    "_rev": "1-cd3074962c936f4b6003a423a445606f",
    "date_of_birth": "1990-03-27",
    "date_of_birth_method": "approx",
    "ephemeral_dob": {
        "age_label": "",
        "age_months": "",
        "age_years": "30",
        "dob_approx": "1990-03-27",
        "dob_iso": "1990-03-27",
        "dob_method": "approx",
        "dob_raw": "1990-03-27",
        "ephemeral_months": "3",
        "ephemeral_years": "1990"
    },
    "external_id": "",
    "meta": {
        "created_by": "chw",
        "created_by_person_uuid": "a7324118-0053-4ce5-8364-e47a0e8f073a",
        "created_by_place_uuid": "f7dd9204-1a35-44fe-899f-5226f8caa509"
    },
    "name": "Test Area Person",
    "notes": "",
    "parent": {
        "_id": "0f97e8e0-48d5-4cd3-a2f5-792a8fa5ca6c",
        "_rev": "1-3fde031c16276eff06a4cf12b59d82bc",
        "contact": {
            "_id": "4a97ae47-ebfe-4d39-8f4e-cf65766f6b80",
            "_rev": "1-cd3074962c936f4b6003a423a445606f",
            "date_of_birth": "1990-03-27",
            "date_of_birth_method": "approx",
            "ephemeral_dob": {
                "age_label": "",
                "age_months": "",
                "age_years": "30",
                "dob_approx": "1990-03-27",
                "dob_iso": "1990-03-27",
                "dob_method": "approx",
                "dob_raw": "1990-03-27",
                "ephemeral_months": "3",
                "ephemeral_years": "1990"
            },
            "external_id": "",
            "meta": {
                "created_by": "chw",
                "created_by_person_uuid": "a7324118-0053-4ce5-8364-e47a0e8f073a",
                "created_by_place_uuid": "f7dd9204-1a35-44fe-899f-5226f8caa509"
            },
            "name": "Test Area Person",
            "notes": "",
            "parent": {
                "_id": "0f97e8e0-48d5-4cd3-a2f5-792a8fa5ca6c",
                "parent": {
                    "_id": "f7dd9204-1a35-44fe-899f-5226f8caa509"
                }
            },
            "phone": "",
            "phone_alternate": "",
            "reported_date": 1585300067473,
            "role": "patient",
            "sex": "female",
            "short_name": "",
            "type": "person"
        },
        "external_id": "",
        "geolocation": "",
        "is_name_generated": "true",
        "meta": {
            "created_by": "chw",
            "created_by_person_uuid": "a7324118-0053-4ce5-8364-e47a0e8f073a",
            "created_by_place_uuid": "f7dd9204-1a35-44fe-899f-5226f8caa509"
        },
        "name": "Test Area Person's Area",
        "notes": "",
        "parent": {
            "_id": "f7dd9204-1a35-44fe-899f-5226f8caa509",
            "_rev": "2-fe0a8f58c1a238b949345a8df97637c4",
            "contact": {
                "_id": "a7324118-0053-4ce5-8364-e47a0e8f073a",
                "_rev": "3-0524407f8a26dec846baac926bc9093f",
                "date_of_birth": "2000-02-01",
                "date_of_birth_method": "",
                "ephemeral_dob": {
                    "dob_approx": "2019-12-12",
                    "dob_calendar": "2000-02-01",
                    "dob_iso": "2000-02-01",
                    "dob_method": "",
                    "dob_raw": "2000-02-01",
                    "ephemeral_months": "12",
                    "ephemeral_years": "2019"
                },
                "external_id": "",
                "meta": {
                    "created_by": "admin",
                    "created_by_person_uuid": "",
                    "created_by_place_uuid": "",
                    "last_edited_by": "admin",
                    "last_edited_by_person_uuid": "",
                    "last_edited_by_place_uuid": ""
                },
                "name": "Health Facility Person",
                "notes": "",
                "parent": {
                    "_id": "f7dd9204-1a35-44fe-899f-5226f8caa509"
                },
                "patient_id": "77306",
                "phone": "+447765902000",
                "phone_alternate": "",
                "reported_date": 1575471568518,
                "role": "chw",
                "sex": "female",
                "short_name": "",
                "type": "person"
            },
            "external_id": "",
            "geolocation": "",
            "is_name_generated": "true",
            "meta": {
                "created_by": "admin",
                "created_by_person_uuid": "",
                "created_by_place_uuid": ""
            },
            "name": "Health Facility Person's Health Facility",
            "notes": "",
            "parent": "",
            "place_id": "99041",
            "reported_date": 1575471568518,
            "type": "district_hospital"
        },
        "reported_date": 1585300067473,
        "type": "health_center"
    },
    "phone": "",
    "phone_alternate": "",
    "reported_date": 1585300067473,
    "role": "patient",
    "sex": "female",
    "short_name": "",
    "type": "person"
}
```

Here is a hydrated report about that same person:

```json
{
    "_attachments": {
        "content": {
            "content_type": "application/xml",
            "digest": "md5-6cHxd65yY4HbLZHFnCutRg==",
            "length": 2807,
            "revpos": 1,
            "stub": true
        }
    },
    "_id": "10aff128-b742-468a-9d09-3837a416673c",
    "_rev": "1-e40ef67a6c80df0cfff1bde85ab48f3c",
    "contact": {
        "_id": "a7324118-0053-4ce5-8364-e47a0e8f073a",
        "_rev": "3-0524407f8a26dec846baac926bc9093f",
        "date_of_birth": "2000-02-01",
        "date_of_birth_method": "",
        "ephemeral_dob": {
            "dob_approx": "2019-12-12",
            "dob_calendar": "2000-02-01",
            "dob_iso": "2000-02-01",
            "dob_method": "",
            "dob_raw": "2000-02-01",
            "ephemeral_months": "12",
            "ephemeral_years": "2019"
        },
        "external_id": "",
        "meta": {
            "created_by": "admin",
            "created_by_person_uuid": "",
            "created_by_place_uuid": "",
            "last_edited_by": "admin",
            "last_edited_by_person_uuid": "",
            "last_edited_by_place_uuid": ""
        },
        "name": "Health Facility Person",
        "notes": "",
        "parent": {
            "_id": "f7dd9204-1a35-44fe-899f-5226f8caa509",
            "_rev": "2-fe0a8f58c1a238b949345a8df97637c4",
            "contact": {
                "_id": "a7324118-0053-4ce5-8364-e47a0e8f073a",
                "_rev": "3-0524407f8a26dec846baac926bc9093f",
                "date_of_birth": "2000-02-01",
                "date_of_birth_method": "",
                "ephemeral_dob": {
                    "dob_approx": "2019-12-12",
                    "dob_calendar": "2000-02-01",
                    "dob_iso": "2000-02-01",
                    "dob_method": "",
                    "dob_raw": "2000-02-01",
                    "ephemeral_months": "12",
                    "ephemeral_years": "2019"
                },
                "external_id": "",
                "meta": {
                    "created_by": "admin",
                    "created_by_person_uuid": "",
                    "created_by_place_uuid": "",
                    "last_edited_by": "admin",
                    "last_edited_by_person_uuid": "",
                    "last_edited_by_place_uuid": ""
                },
                "name": "Health Facility Person",
                "notes": "",
                "parent": {
                    "_id": "f7dd9204-1a35-44fe-899f-5226f8caa509"
                },
                "patient_id": "77306",
                "phone": "+447765902000",
                "phone_alternate": "",
                "reported_date": 1575471568518,
                "role": "chw",
                "sex": "female",
                "short_name": "",
                "type": "person"
            },
            "external_id": "",
            "geolocation": "",
            "is_name_generated": "true",
            "meta": {
                "created_by": "admin",
                "created_by_person_uuid": "",
                "created_by_place_uuid": ""
            },
            "name": "Health Facility Person's Health Facility",
            "notes": "",
            "parent": "",
            "place_id": "99041",
            "reported_date": 1575471568518,
            "type": "district_hospital"
        },
        "patient_id": "77306",
        "phone": "+447765902000",
        "phone_alternate": "",
        "reported_date": 1575471568518,
        "role": "chw",
        "sex": "female",
        "short_name": "",
        "type": "person"
    },
    "content_type": "xml",
    "fields": {
        "data": {
            "__date_of_death": "2020-03-27",
            "__death_information": "",
            "__place_of_death": "health_facility",
            "__place_of_death_other": "",
            "meta": {
                "__household_uuid": "",
                "__patient_id": "",
                "__patient_uuid": "4a97ae47-ebfe-4d39-8f4e-cf65766f6b80",
                "__source": "contact",
                "__source_id": ""
            }
        },
        "death_details": {
            "date_of_death": "2020-03-27",
            "death_information": "",
            "place_of_death": "health_facility"
        },
        "group_review": {
            "blank_note": "",
            "c_patient_age": "30 years old",
            "r_death_info": "",
            "r_key_instruction": "",
            "r_patient_details": "",
            "r_referral": "",
            "r_summary_details": "",
            "r_undo": "",
            "submit": ""
        },
        "inputs": {
            "contact": {
                "_id": "4a97ae47-ebfe-4d39-8f4e-cf65766f6b80",
                "date_of_birth": "1990-03-27",
                "name": "Test Area Person",
                "parent": {
                    "parent": {
                        "contact": {
                            "name": "Health Facility Person",
                            "phone": "+447765902000"
                        }
                    }
                },
                "patient_id": "",
                "sex": "female",
                "short_name": ""
            },
            "meta": {
                "deprecatedID": "",
                "location": {
                    "error": "",
                    "lat": "",
                    "long": "",
                    "message": ""
                }
            },
            "source": "contact",
            "source_id": ""
        },
        "meta": {
            "instanceID": "uuid:6577675b-2b0c-45d3-9db0-8cf2903eb946"
        },
        "patient_age_in_days": "10958",
        "patient_age_in_months": "360",
        "patient_age_in_years": "30",
        "patient_display_name": "Test Area Person",
        "patient_id": "",
        "patient_name": "Test Area Person",
        "patient_short_name": "",
        "patient_uuid": "4a97ae47-ebfe-4d39-8f4e-cf65766f6b80"
    },
    "form": "death_report",
    "from": "+447765902000",
    "hidden_fields": [
        "patient_age_in_years",
        "patient_age_in_months",
        "patient_age_in_days",
        "patient_uuid",
        "patient_id",
        "patient_name",
        "patient_short_name",
        "patient_display_name",
        "group_review",
        "data",
        "meta"
    ],
    "patient": {
        "_id": "4a97ae47-ebfe-4d39-8f4e-cf65766f6b80",
        "_rev": "1-cd3074962c936f4b6003a423a445606f",
        "date_of_birth": "1990-03-27",
        "date_of_birth_method": "approx",
        "ephemeral_dob": {
            "age_label": "",
            "age_months": "",
            "age_years": "30",
            "dob_approx": "1990-03-27",
            "dob_iso": "1990-03-27",
            "dob_method": "approx",
            "dob_raw": "1990-03-27",
            "ephemeral_months": "3",
            "ephemeral_years": "1990"
        },
        "external_id": "",
        "meta": {
            "created_by": "chw",
            "created_by_person_uuid": "a7324118-0053-4ce5-8364-e47a0e8f073a",
            "created_by_place_uuid": "f7dd9204-1a35-44fe-899f-5226f8caa509"
        },
        "name": "Test Area Person",
        "notes": "",
        "parent": {
            "_id": "0f97e8e0-48d5-4cd3-a2f5-792a8fa5ca6c",
            "_rev": "1-3fde031c16276eff06a4cf12b59d82bc",
            "contact": {
                "_id": "4a97ae47-ebfe-4d39-8f4e-cf65766f6b80"
            },
            "external_id": "",
            "geolocation": "",
            "is_name_generated": "true",
            "meta": {
                "created_by": "chw",
                "created_by_person_uuid": "a7324118-0053-4ce5-8364-e47a0e8f073a",
                "created_by_place_uuid": "f7dd9204-1a35-44fe-899f-5226f8caa509"
            },
            "name": "Test Area Person's Area",
            "notes": "",
            "parent": {
                "_id": "f7dd9204-1a35-44fe-899f-5226f8caa509",
                "_rev": "2-fe0a8f58c1a238b949345a8df97637c4",
                "contact": {
                    "_id": "a7324118-0053-4ce5-8364-e47a0e8f073a",
                    "parent": {
                        "_id": "f7dd9204-1a35-44fe-899f-5226f8caa509"
                    }
                },
                "external_id": "",
                "geolocation": "",
                "is_name_generated": "true",
                "meta": {
                    "created_by": "admin",
                    "created_by_person_uuid": "",
                    "created_by_place_uuid": ""
                },
                "name": "Health Facility Person's Health Facility",
                "notes": "",
                "parent": "",
                "place_id": "99041",
                "reported_date": 1575471568518,
                "type": "district_hospital"
            },
            "reported_date": 1585300067473,
            "type": "health_center"
        },
        "phone": "",
        "phone_alternate": "",
        "reported_date": 1585300067473,
        "role": "patient",
        "sex": "female",
        "short_name": "",
        "type": "person"
    },
    "reported_date": 1585300079150,
    "type": "data_record"
}
```

**Comment by garethbowen** [comment]
> doc.patient, all of those patient's parents, but not those parent's contacts (is this a bug @garethbowen ?)

I'd say so, please raise it, as well as a documentation bug for communicating this (especially helpful with the upcoming hydration API).

File: api/src/controllers/hydration.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
+const db = require('../db');
+const lineageFactory = require('@medic/lineage');
+const serverUtils = require('../server-utils');
+
+const lineage = lineageFactory(Promise, db.medic);
+
+module.exports = {
+  hydrate: (req, res) => {
+    const docIds = Array.isArray(req.body) ? req.body : [req.body];
```
Comment (SCdF) [inline api/src/controllers/hydration.js:9]: TODO: check permissions (otherwise this exposes all data to all people)

**Comment by dianabarsan** [comment]
> doc.patient, all of those patient's parents, but not those parent's contacts (is this a bug?)

Confirming this is a bug that only exists in the `fetchHydratedDoc` function. `hydrateDocs` correctly hydrates report.patient.parent.contact. I'm in the process of fixing this.

**Comment by garethbowen** [comment]
The full RapidPro integration has been merged so I assume this is no longer needed.

**Pull request closed without merging**